### DISCUSSION TITLE
Does the tree (or a node) store the original source text?

### DISCUSSION THREAD

**Discussion by leonerd** [tree-sitter/tree-sitter#2023] [Q&A] [answered]
When you have an anonymous node (such as representing a bracket or operator symbol), its name is the text from the source code. But when you have a named one such as represents a string or number literal or identifier, it does not. Is the original source text stored anywhere?

I'm asking as I'd like to provide a "text substring" method on the tree node objects in my language binding. If such isn't stored, I guess it will have to be provided by some sort of higher-level wrapping on top.

How do language bindings generally handle this sort of thing?

**Comment by ahelwer** [comment, marked as answer]
No, the text is not stored in the tree. Nodes store byte offsets you can use to index into the text. The consumer of the grammar will always have access to the text, otherwise how else would they get a parse tree in the first place?
